### PR TITLE
Support for parsing up PARTITION table ddl

### PR DESCRIPTION
**Pull request by osheroff**
we use the same trickery we used in parsing the CHECK() function; we don't want to bother parsing a complex mysql expression, so we hint to the parser that we want to skip a matching set of parenthesis, and the parser does a token-stream rewrite and starts the parse over again.

Note that bringing in the partition sql (generated out of the mysql-server test suite) also exposed some issues in the string literal tokenizer.

@zendesk/rules

### CODE REVIEW

**Comment by osheroff** [comment]
addresses issue #387

File: src/main/antlr4/imports/mysql_partition.g4 [hunk 1]
```diff
@@ -0,0 +1,68 @@
+grammar mysql_partition;
+
+import mysql_idents;
+
+partition_by:
+  PARTITION BY
+  partition_by_what
+  partition_count?
+  subpartition_by?
```
Comment (dadah89) [inline src/main/antlr4/imports/mysql_partition.g4]: missing `[SUBPARTITIONS num]`?

Comment (dadah89) [inline src/main/antlr4/imports/mysql_partition.g4]: oh nvm, part of `subpartition_by`

File: src/main/antlr4/imports/mysql_partition.g4 [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  ;
+
+partition_by_what:
+    LINEAR? HASH skip_parens
+  | LINEAR? KEY (ALGORITHM '=' INTEGER_LITERAL)? skip_parens
+  | RANGE COLUMNS? skip_parens
+  | LIST COLUMNS? skip_parens;
+
+subpartition_by:
+  SUBPARTITION BY partition_by_what (SUBPARTITIONS INTEGER_LITERAL)?;
```
Comment (osheroff) [inline src/main/antlr4/imports/mysql_partition.g4]: @dadah89 the nubmer of subpartitions is here.

File: src/main/antlr4/imports/mysql_partition.g4 [hunk 3]
```diff
@@ -0,0 +1,68 @@
+grammar mysql_partition;
+
+import mysql_idents;
+
+partition_by:
+  PARTITION BY
+  partition_by_what
+  partition_count?
+  subpartition_by?
+  partition_definitions?
```
Comment (dadah89) [inline src/main/antlr4/imports/mysql_partition.g4]: `skip_parens` doesn't work here? guess it could mach something that it shouldn't?

Comment (osheroff) [inline src/main/antlr4/imports/mysql_partition.g4]: it.. might, if it was `skip_parens?`  generally I've tried to only use the skip-parens code when I'm *sure* there's parenthesis to be skipped (ie they're anchored to another expression).  The whole partition definition stuff here is optional, so yeah, could maybe be something else.

**Comment by dadah89** [comment]
lgtm 👍

**Pull request merged**